pkg/controller/mssql/grant/reconciler.go

```diff
@@ -0,0 +1,275 @@
+		managed.WithLogger(l.WithValues("controller", name)),
+		managed.WithPollInterval(10*time.Minute),
+		managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name))))
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.Grant{}).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: maxConcurrency,
+		}).
+		Complete(r)
+}
+
+type connector struct {
+	kube  client.Client
+	usage resource.Tracker
+	newDB func(creds map[string][]byte, database string) xsql.DB
+}
+
+func (c *connector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	cr, ok := mg.(*v1alpha1.Grant)
+	if !ok {
+		return nil, errors.New(errNotGrant)
+	}
+
+	if err := c.usage.Track(ctx, mg); err != nil {
+		return nil, errors.Wrap(err, errTrackPCUsage)
+	}
+
+	// ProviderConfigReference could theoretically be nil, but in practice the
+	// DefaultProviderConfig initializer will set it before we get here.
+	pc := &v1alpha1.ProviderConfig{}
+	if err := c.kube.Get(ctx, types.NamespacedName{Name: cr.GetProviderConfigReference().Name}, pc); err != nil {
+		return nil, errors.Wrap(err, errGetPC)
+	}
+
+	// We don't need to check the credentials source because we currently only
+	// support one source (MSSQLConnectionSecret), which is required and
+	// enforced by the ProviderConfig schema.
+	ref := pc.Spec.Credentials.ConnectionSecretRef
+	if ref == nil {
+		return nil, errors.New(errNoSecretRef)
+	}
+
+	s := &corev1.Secret{}
+	if err := c.kube.Get(ctx, types.NamespacedName{Namespace: ref.Namespace, Name: ref.Name}, s); err != nil {
+		return nil, errors.Wrap(err, errGetSecret)
+	}
+
+	return &external{
+		db:   c.newDB(s.Data, stringValue(cr.Spec.ForProvider.Database)),
+		kube: c.kube,
+	}, nil
+}
+
+type external struct {
+	db   xsql.DB
+	kube client.Client
+}
+
+func (c *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.Grant)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotGrant)
+	}
+
+	permissions, err := c.getPermissions(ctx, *cr.Spec.ForProvider.User)
+	if err != nil {
+		return managed.ExternalObservation{}, err
+	}
+	if len(permissions) == 0 {
+		return managed.ExternalObservation{}, nil
+	}
+
+	cr.SetConditions(xpv1.Available())
+
+	g, r := diffPermissions(cr.Spec.ForProvider.Permissions.ToStringSlice(), permissions)
+	return managed.ExternalObservation{
+		ResourceExists:   true,
+		ResourceUpToDate: len(g) == 0 && len(r) == 0,
+	}, nil
+}
+
+func (c *external) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*v1alpha1.Grant)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotGrant)
+	}
+
+	username := *cr.Spec.ForProvider.User
+	permissions := strings.Join(cr.Spec.ForProvider.Permissions.ToStringSlice(), ", ")
+
+	query := fmt.Sprintf("GRANT %s TO %s", permissions, mssql.QuoteIdentifier(username))
```

Comment:
Don't we need to quote the permission values?

Reply:
Looks like not: https://docs.microsoft.com/en-us/sql/t-sql/statements/grant-database-permissions-transact-sql?view=sql-server-ver15#a-granting-permission-to-create-tables